core/arch/arm/kernel/delay.c

```diff
@@ -44,3 +48,37 @@ void mdelay(uint32_t ms)
 {
 	udelay(1000 * ms);
 }
+
+uint64_t utimeout_init(uint32_t us)
+{
+	return read_cntpct() + US2CNT(us);
+}
+
+bool utimeout_elapsed(uint32_t us, uint64_t reference)
+{
+	uint64_t origin = reference - US2CNT(us);
+	uint64_t now = read_cntpct();
+
+	if (origin < reference)
+		return now < origin || now > reference;
+
+	return now < origin && now > reference;
+}
+
+unsigned int utimeout_elapsed_us(uint32_t us, uint64_t reference)
```

Comment:
I think the return type should be `uint32_t` instead as that's what we use as input

Reply:
You're tight. I think using an `unsigned int` as input is even better.

Reply:
That depends on which range of values we're aiming for. An advantage with `uint32_t` is that the range is well defined, while `unsigned int` is just some moderately large unsigned integer.

Reply:
`udelay() and `mdelay()` use `uint32_t` so this type is better for consistency at least

Reply:
ok. agreed.